src/client/producer.rs

```diff
@@ -323,64 +326,246 @@ impl ProducerClient for PartitionClient {
     }
 }
 
-/// [`BatchProducer`] attempts to aggregate multiple produce requests together
-/// using the provided [`Aggregator`].
-///
-/// It will buffer up records until either the linger time expires or [`Aggregator`]
-/// cannot accommodate another record.
-///
-/// At this point it will flush the [`Aggregator`]
-///
-/// [`Aggregator`]: aggregator::Aggregator
 #[derive(Debug)]
-pub struct BatchProducer<A>
+struct ProducerInner<A>
 where
     A: aggregator::Aggregator,
 {
-    linger: Duration,
+    /// A wrapper over an [`Aggregator`] implementation that represents a single
+    /// aggregated batch of Records.
+    ///
+    /// This is wrapped in an [`Option`] to enable the
+    /// [`ProducerInner::flush()`] call to take the buffer, flush it, and
+    /// replace it with a new instance. Only the `flush()` method will ever
+    /// observe this being `None`, and all other code can safely `unwrap()` on
+    /// it.
+    batch_builder: Option<BatchBuilder<A>>,
+
+    /// A logical clock to enable a conditional flush of a specific
+    /// [`BatchBuilder`] instance.
+    ///
+    /// Each time a new [`BatchBuilder`] is initialised, this counter increases
+    /// by 1 (and wraps as necessary). Callers requested to flush the batch
+    /// after the linger duration are given the current value of this LC in the
+    /// [`CallerRole::Linger`] return variant, and once the linger fires, they
+    /// present their LC value in the [`ProducerInner::flush()`] call. If the
+    /// values match, the buffer if flushed. If the values do not match, the
+    /// buffer the caller was responsible for has already been flushed, and the
+    /// flush call becomes a NOP.
+    flush_clock: usize,
+
+    /// Used to track if a [`CallerRole::Linger`] has been returned for the
+    /// current batch_builder.
+    has_linger_waiter: bool,
 
     compression: Compression,
-
     client: Arc<dyn ProducerClient>,
 
-    inner: Arc<Mutex<ProducerInner<A>>>,
+    /// A list of (potentially) outstanding flush tasks.
+    ///
+    /// These may or may not yet be complete, and completed flush tasks are
+    /// removed from this list when adding new flush tasks or manually flushing
+    /// with a call to [`BatchProducer::flush()`].
+    pending_flushes: Vec<JoinHandle<()>>,
```

Comment:
We should abort all background tasks if all user callsides are gone or cancelled:

```rust
impl Drop for ProducerInner {
    fn drop(&mut self) {
        for handle in self.pending_flushes.drain(..) {
            handle.abort();
        }
    }
}
```

Technically I'm wondering if we should even cancel a specific flush tasks if all callsides for that specific task are gone :thinking: Because what's the point of sending this data then.

BTW: You may assume that the `PartitionClient` is cancellation-safe (everything else would be a bug). The underlying `Messenger` ensure that we never send partial messages to the broker.

Reply:
c0ba668